### ISSUE TITLE
Plugin-content-pages preset options, 'routeBasePath' can't be ''

### ISSUE THREAD

**Issue by Niko030303** [facebook/docusaurus#8254]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Description

When I use default preset options
```shell
module.exports = {
  presets: [
    [
      '@docusaurus/preset-classic',
      {
        pages: {
          path: 'src/pages',
          routeBasePath: '',
          include: ['**/*.{js,jsx,ts,tsx,md,mdx}'],
          exclude: [
            '**/_*.{js,jsx,ts,tsx,md,mdx}',
            '**/_*/**',
            '**/*.test.{js,jsx,ts,tsx}',
            '**/__tests__/**',
          ],
          mdxPageComponent: '@theme/MDXPage',
          remarkPlugins: [require('remark-math')],
          rehypePlugins: [],
          beforeDefaultRemarkPlugins: [],
          beforeDefaultRehypePlugins: [],
        },
      },
    ],
  ],
};
```
It pointed out that `ValidationError: "routeBasePath" is not allowed to be empty`
![image](https://user-images.githubusercontent.com/82041918/198238848-6cbca559-943a-471b-a608-5490716014cf.png)

### Self-service

- [X] I'd be willing to address this documentation request myself.

**Comment by slorber** [comment]
Hey

Somehow it's done on purpose because previously it lead to issues, although now with https://github.com/facebook/docusaurus/pull/3427 it shouldn't be an issue anymore.

We could either:
- forbid '' but print a better error message suggesting to use "/" instead
- allow '' to pass validation, as it should work fine now

Considering we have related validation for baseUrl (https://github.com/facebook/docusaurus/pull/8066) I guess we should do the same here, and apply this to all 3 content plugins: docs, pages, blog.

To make things more predictable I'd prefer if we apply the same logic as baseUrl and always normalize this URL segment to have leading/trailing slashes. 

Cf https://github.com/facebook/docusaurus/pull/8066 for implementation

```js
  baseUrl: Joi.string()
    .required()
    .custom((value: string) => addLeadingSlash(addTrailingSlash(value))),
```

Any opinion @Josh-Cena ?

---

Good issue for an external contributor looking to contribute. 

Send directly the PR only if you can handle it. Do not claim it, we won't assign it to you unless there's a PR first.

**Comment by Djunnni** [comment]
@slorber 

could you check my PR? i reflect your opinion

> Hey
> 
> Somehow it's done on purpose because previously it lead to issues, although now with #3427 it shouldn't be an issue anymore.
> 
> We could either:
> 
> * forbid '' but print a better error message suggesting to use "/" instead
> * allow '' to pass validation, as it should work fine now
> 
> Considering we have related validation for baseUrl (#8066) I guess we should do the same here, and apply this to all 3 content plugins: docs, pages, blog.
> 
> To make things more predictable I'd prefer if we apply the same logic as baseUrl and always normalize this URL segment to have leading/trailing slashes.
> 
> Cf #8066 for implementation
> 
> ```js
>   baseUrl: Joi.string()
>     .required()
>     .custom((value: string) => addLeadingSlash(addTrailingSlash(value))),
> ```
>

**Comment by Josh-Cena** [comment]
Frankly I don't get the semantic difference between `"/"` and `""`. Is the latter only useful to hack around the "multiple root plugins" issue? If that's the case I think we should continue to reject it until the actual issue is properly remedied. We already normalize slashes, but I don't think we should allow empty strings unless it has clear semantics.

**Comment by Djunnni** [comment]
> Frankly I don't get the semantic difference between `"/"` and `""`. Is the latter only useful to hack around the "multiple root plugins" issue? If that's the case I think we should continue to reject it until the actual issue is properly remedied. We already normalize slashes, but I don't think we should allow empty strings unless it has clear semantics.

I agree with your opinion, but here, even when routeBasePath is not defined, the root (/) path is entered by default. I'm not quite sure what the difference is with "" in this case.
I think it would be better to put it as an essential element rather than to remove the ambiguous case.

**Comment by slorber** [comment]
> Frankly I don't get the semantic difference between `'/'` and `''`. Is the latter only useful to hack around the "multiple root plugins" issue?

I don't even remember what this issue is exactly but I don't think it hacks around anything if we normalize '' to '/' anyway.

> If that's the case I think we should continue to reject it until the actual issue is properly remedied. 

'/' works so if we normalize '' to '/' there won't be issues, it would work as if user provided "/" 🤷‍♂️ 


> We already normalize slashes, but I don't think we should allow empty strings unless it has clear semantics.

To me using "docs" (default plugin routeBasePath value) instead of "/docs/" is not particularly more semantically wrong than using '' instead of '/'.

---


`routeBasePath` is somehow the "plugin baseUrl". It made sense to me to be consistent between `baseUrl` and `routeBasePath`

Now I didn't notice but we actually don't accept `baseUrl: ''` . I thought we did normalize it to "/" already, but instead it prints an error message:

> "baseUrl" is not allowed to be empty

I suggest either:
- allow '' for both `baseUrl` and `routeBasePath`, eventually emit a warning saying the value should rather be '/'
- reject '' for both `baseUrl` and `routeBasePath` with a clearer error message explaining to use '/' instead (+ fix docs using '')

The first option looks better to me, as users will intuitively try to use ''. 

Do we want to throw on start while we could handle the bad value more gracefully?

**Comment by Djunnni** [comment]
If the progress is stopped, how about we discuss it again?

* allow '' for both baseUrl and routeBasePath, eventually emit a warning saying the value should rather be '/'

how about think that @Josh-Cena ?

**Comment by Josh-Cena** [comment]
> `routeBasePath` is somehow the "plugin baseUrl". It made sense to me to be consistent between `baseUrl` and `routeBasePath`

Yeah, I'll agree with that stance. Well, I'll prefer to allow it, but normalize—I know it was previously used as a "multiple root plugins" hack, which is why I don't like `""`; but if we always normalize, I do prefer the non-erroring side.

**Comment by Djunnni** [comment]
> I know it was previously used as a "multiple root plugins" hack, which is why I don't like ""

is it happen in this repository? if some of case is exist. i want to know more. because i didn't hear vulnerabilities.

i readed [docusaurus routing](https://docusaurus.io/docs/advanced/routing)

> `routeBasePath` is "plugin baseUrl". 

i think it can be used "" because it works in root("/")
